relay-general/src/store/transactions/rules.rs

```diff
@@ -0,0 +1,195 @@
+use chrono::{DateTime, Utc};
+use relay_common::Glob;
+use serde::{Deserialize, Deserializer, Serialize};
+
+use crate::protocol::TransactionSource;
+
+/// Helper function to deserialize the string patter into the [`relay_common::Glob`].
+fn deserialize_glob_pattern<'de, D>(deserializer: D) -> Result<Glob, D::Error>
+where
+    D: Deserializer<'de>,
+{
+    let pattern = String::deserialize(deserializer)?;
+    let glob = Glob::builder(&pattern)
+        .capture_star(true)
+        .capture_double_star(false)
+        .capture_question_mark(false)
+        .build();
+    Ok(glob)
+}
+
+/// Default value for substitution in [`RedactionRule`].
+fn default_substitution() -> String {
+    "*".to_string()
+}
+
+/// Contains transaction attribute the rule must only be applied to.
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+pub struct RuleScope {
+    /// The source of the transaction.
+    pub source: TransactionSource,
+}
+
+impl Default for RuleScope {
+    fn default() -> Self {
+        Self {
+            source: TransactionSource::Url,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+pub struct Replace {
+    #[serde(default = "default_substitution")]
+    substitution: String,
+}
+
+/// Describes what to do with the matched pattern.
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+#[serde(tag = "method", rename_all = "snake_case")]
+pub enum RedactionRule {
+    Replace(Replace),
+    #[serde(other, skip_serializing)]
+    Other,
+}
+
+impl Default for RedactionRule {
+    fn default() -> Self {
+        Self::Replace(Replace {
+            substitution: default_substitution(),
+        })
+    }
+}
+
+/// The rule describes how transaction name should be changed.
+#[derive(Debug, Clone, Serialize, Deserialize, Eq, PartialEq)]
+pub struct TransactionNameRule {
+    /// The pattern which will be applied to transaction name.
+    #[serde(deserialize_with = "deserialize_glob_pattern")]
+    pub pattern: Glob,
+    /// Date time when the rule expires and it should not be applied anymore.
+    pub expiry: DateTime<Utc>,
+    /// Object containing transaction attributes the rules must only be applied to.
+    #[serde(default)]
+    pub scope: RuleScope,
+    /// Object describing what to do with the matched pattern.
+    pub redaction: RedactionRule,
+}
+
+impl TransactionNameRule {
+    /// Applies the rule to the provided value.
+    ///
+    /// Note: currently only `url` source for rules supported.
+    pub fn apply(&self, value: &str) -> Option<String> {
+        match &self.redaction {
+            RedactionRule::Replace(Replace { substitution }) => {
+                let result = self.pattern.replace_captures(value, substitution);
+                Some(result)
+            }
+            _ => {
+                relay_log::trace!("Replacement rule type is unsupported!");
+                None
```

Comment:
The only reason this can return `None` is because of the `#[serde(other)]` option on `RedactionRule`, right? I would instead change the signature to `-> String`, and return `value.to_owned()` here.

Ideally, we would remove unsupported replacement rules on deserialization. Then this branch practically becomes dead code.

Reply:
Or, if we add a `matches` function as suggested below, that could automatically return false for any unsupported rule.